From 74543aa807ad8afb64ced8439852e15fbf5ee8f5 Mon Sep 17 00:00:00 2001
From: Suresh Gupta <suresh.gupta@xilinx.com>
Date: Fri, 5 Oct 2018 19:42:55 +0530
Subject: [PATCH 0448/1852] v4l: xilinx-multi-scaler: Corrects the format
 struct in TRY_FMT

commit 5b6179c234e2edb78be0e32949539a2c907cd8c2 from
https://github.com/Xilinx/linux-xlnx.git

V4L2 specification suggests the driver corrects the
format struct if any of the dimensions is unsupported.

The patch limits the minimum height and width to 64 and
maximum height and width to the value set in DT.

Signed-off-by: Suresh Gupta <suresh.gupta@xilinx.com>
Reviewed-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../platform/xilinx/xilinx-multi-scaler.c     | 23 ++++++++++++++-----
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-multi-scaler.c b/drivers/media/platform/xilinx/xilinx-multi-scaler.c
index e95a1acd4fa8..66c322eba0fb 100644
--- a/drivers/media/platform/xilinx/xilinx-multi-scaler.c
+++ b/drivers/media/platform/xilinx/xilinx-multi-scaler.c
@@ -1065,12 +1065,23 @@ vidioc_try_fmt(struct xm2msc_chan_ctx *chan_ctx, struct v4l2_format *f)
 	int index;
 
 	if (pix->width < XM2MSC_MIN_WIDTH || pix->width > xm2msc->max_wd ||
-	    pix->height < XM2MSC_MIN_HEIGHT || pix->height > xm2msc->max_ht) {
-		v4l2_err(&xm2msc->v4l2_dev,
-			 "Wrong input parameters %d, wxh: %dx%d.\n",
-			 f->type, f->fmt.pix.width, f->fmt.pix.height);
-		return -EINVAL;
-	}
+	    pix->height < XM2MSC_MIN_HEIGHT || pix->height > xm2msc->max_ht)
+		dev_dbg(xm2msc->dev,
+			"Wrong input parameters %d, wxh: %dx%d.\n",
+			f->type, f->fmt.pix.width, f->fmt.pix.height);
+	/*
+	 * V4L2 specification suggests the driver corrects the
+	 * format struct if any of the dimensions is unsupported
+	 */
+	if (pix->height < XM2MSC_MIN_HEIGHT)
+		pix->height = XM2MSC_MIN_HEIGHT;
+	else if (pix->height > xm2msc->max_ht)
+		pix->height = xm2msc->max_ht;
+
+	if (pix->width < XM2MSC_MIN_WIDTH)
+		pix->width = XM2MSC_MIN_WIDTH;
+	else if (pix->width > xm2msc->max_wd)
+		pix->width = xm2msc->max_wd;
 
 	vq = v4l2_m2m_get_vq(chan_ctx->m2m_ctx, f->type);
 	if (!vq)
-- 
2.31.1

